### PR TITLE
dev/core#61 Split Edit Message Templates Permission

### PR DESCRIPTION
**Pull request by ajesamson**
Overview
----------------------------------------
The `edit message templates` permission currently grants access to user-driven messages and system workflow message. The implication of this is that, there is no way of restricting a particular user to just user-driven messages or system workflow messages. 

This PR adds 2 new granular permissions called `edit user-driven message templates` and `edit system workflow message templates`. This helps to further limit which of the templates a user can view and edit, while retaining exiting `edit message templates` permission for backward compability.

Before
----------------------------------------
before

After
----------------------------------------
after

Technical Details
----------------------------------------
New permissions were added to [existing ones](https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/Permission.php)
```php
      'edit system workflow message templates' => array(
        $prefix . ts('edit system workflow message templates'),
      ),
      'edit user-driven message templates' => array(
        $prefix . ts('edit user-driven message templates'),
      ),
```
and applied to [message template urls](https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/xml/Menu/Admin.xml#L258).
```php
<item>
     <path>civicrm/admin/messageTemplates</path>
     <title>Message Templates</title>
     <desc>Message templates allow you to save and re-use messages with layouts which you can use when sending email to one or more contacts.</desc>
     <page_callback>CRM_Admin_Page_MessageTemplates</page_callback>
     <adminGroup>Communications</adminGroup>
     <icon>admin/small/template.png</icon>
     <access_arguments>edit message templates;edit user-driven message templates;edit system workflow message templates</access_arguments>
     <weight>30</weight>
  </item>
```
Permission checks were applied on the [view page](https://github.com/civicrm/civicrm-core/blob/master/templates/CRM/Admin/Page/MessageTemplates.tpl), restricting tab access. Checks were also carried out while [building the form](https://github.com/civicrm/civicrm-core/blob/master/CRM/Admin/Form/MessageTemplates.php#L91) to ensure users with one permission, does not edit forms belonging to other permission group.
```php
  private function checkUserPermission($workflowId) {
    if (isset($workflowId)) {
      $canView = CRM_Core_Permission::check('edit system workflow message templates');
    } else {
      $canView = CRM_Core_Permission::check('edit user-driven message templates');
    }

    if (! $canView && ! CRM_Core_Permission::check('edit message templates')) {
      CRM_Core_Session::setStatus(ts('You do not have permission to view requested page.'), ts('Access Denied'));
      $url = CRM_Utils_System::url('civicrm/admin/messageTemplates', "reset=1");
      CRM_Utils_System::redirect($url);
    }
  }
```
For existing users to be aware of these changes, a pre-upgrade message notification was provided. This affords them the opportunity of deciding whether to upgrade existing user permissions or not.
```php
if ($rev == '5.2.alpha1') {
  $params = array(
    1 => 'edit user-driven message templates',
    2 => 'edit system workflow message templates',
    3 => 'edit message templates',
  );
  $preUpgradeMessage .= '<p>' . ts('New granular permissions called %1 and %2 have been added for %3 permission. These permissions help to limit user access per template', $params) . '</p>';
}
```

### CODE REVIEW

**Comment by seamuslee001** [comment]
Jenkins ok to test

**Comment by eileenmcnaughton** [comment]
@mattwire perhaps you can trade with Compucorp to review one of yours? I know you want review on your membership tidy up one. 

I think the idea of splitting up the permissions makes sense in principle. I'm a bit alarmed to see those php calls from the tpl file but I'm sure there are ample other places where we check permissions to compare to. If we really don't have an existing smarty permission check maybe we should add one in CRM_Core_Smarty_Plugins?

File: CRM/Core/xml/Menu/Admin.xml [hunk 1]
```diff
@@ -262,15 +262,15 @@
      <page_callback>CRM_Admin_Page_MessageTemplates</page_callback>
      <adminGroup>Communications</adminGroup>
      <icon>admin/small/template.png</icon>
-     <access_arguments>edit message templates</access_arguments>
+     <access_arguments>edit message templates;edit user-driven message templates;edit system workflow message templates</access_arguments>
```
Comment (eileenmcnaughton) [inline CRM/Core/xml/Menu/Admin.xml]: I see the xml sometimes uses commas & sometimes semi-colons - do they have the same meaning?

Comment (seamuslee001) [reply CRM/Core/xml/Menu/Admin.xml]: I think it’s the diff between or and I think ; == or , == and

Comment (mickadoo) [reply CRM/Core/xml/Menu/Admin.xml]: I think Seamus is [right](https://docs.civicrm.org/dev/en/latest/framework/routing/#access_arguments)

File: templates/CRM/Admin/Page/MessageTemplates.tpl [hunk 2]
```diff
@@ -89,14 +89,23 @@
 <div class="crm-content-block crm-block">
   <div id='mainTabContainer'>
     <ul>
-      <li id='tab_user'>    <a href='#user'     title='{ts}User-driven Messages{/ts}'>    {ts}User-driven Messages{/ts}    </a></li>
-      <li id='tab_workflow'><a href='#workflow' title='{ts}System Workflow Messages{/ts}'>{ts}System Workflow Messages{/ts}</a></li>
+      {if $canEditUserDrivenMessageTemplates or $canEditMessageTemplates}
+        <li id='tab_user'>    <a href='#user'     title='{ts}User-driven Messages{/ts}'>    {ts}User-driven Messages{/ts}    </a></li>
```
Comment (mickadoo) [inline templates/CRM/Admin/Page/MessageTemplates.tpl]: I know it wasn't something you introduced, but can you remove the extra spaces here like in the second link below?

File: CRM/Admin/Form/MessageTemplates.php [hunk 3]
```diff
@@ -214,6 +215,26 @@ public function buildQuickForm() {
     }
   }
 
+  /**
+   * Restrict users access based on permission
+   *
+   * @param int $workflowId
+   */
+  private function checkUserPermission($workflowId) {
+    if (isset($workflowId)) {
```
Comment (mickadoo) [inline CRM/Admin/Form/MessageTemplates.php]: I'm not really familiar with message templates so just a question here, is it impossible for user driven message templates to have a workflow ID?

**Review by mickadoo — CHANGES_REQUESTED** [review]
looks pretty good overall, some minor comments

File: CRM/Admin/Form/MessageTemplates.php (hunk 3 above)
Comment (ajesamson) [reply CRM/Admin/Form/MessageTemplates.php]: User driven message templates do not have workflow ID. A template becomes system workflow message template the moment it has workflow ID.

File: templates/CRM/Admin/Page/MessageTemplates.tpl (hunk 2 above)
Comment (ajesamson) [reply templates/CRM/Admin/Page/MessageTemplates.tpl]: Alright i will.

**Review by mickadoo — APPROVED** [review]
Looks ok to me, it'll require someone (non compucorp) to review too though

**Comment by mickadoo** [comment]
* Explain ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain))
    * __PASS__ : The goal/problem/solution have been adequately explained in the PR.
* Test results ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test))
    * __UNREVIEWED__
* Code quality ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code))
    * __PASS__: The functionality, purpose, and style of the code seems clear+sensible.
* Documentation ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc))
    * __UNREVIEWED__
    * __COMMENTS__: Do you think this needs a documentation change @ajesamson?
* Maintainability ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint))
    * __UNREVIEWED__
* Run it ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run))
    * __UNREVIEWED__
* User impact ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user))
    * __UNREVIEWED__

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by colemanw** [comment]
Hi @ajesamson  thanks for this contribution. I'm just looking at it for the first time and my initial question is about whether the original permission indeed ought to be kept around. That seems needlessly confusing, as it will have significant overlap with the new permissions.

**Comment by eileenmcnaughton** [comment]
@colemanw ah yeah - you'd need to add in a capture I suppose which starts to get a bit complex

**Comment by ajesamson** [comment]
Hi @colemanw the reason why it was left behind is more or less for backward compatibility. It was meant to allow existing users with permission continue working without impacting their process on upgrade. Your suggestions are highly welcomed on how best to handle this.

**Comment by mattwire** [comment]
@ajesamson I would have thought we could remove the old permission, making sure we add an upgrader message (there was one when batch permissions were added) telling the user that they will need to reconfigure permissions for message templates.  The only other place it may cause a problem is if that permission is used in extensions, @totten normally says "grep universe" to find out.

**Comment by colemanw** [comment]
I agree. As part of the upgrade, I think all users with the old permission should be granted the 2 new ones. And the upgrade message can give admins a heads up that they now have finer control if they want it.

**Comment by ajesamson** [comment]
@colemanw Please can you provide a guide on how best to migrate the old permission for existing users to 2 new ones. I have spent some time looking through the code base but could not get a method that provides easy access to CMS permission table for such upgrade.

The [suggested sample code](https://github.com/civicrm/civicrm-core/blob/master/CRM/Upgrade/Incremental/php/FourSeven.php#L75) by @mattwire used `setPreUpgradeMessage` and left decision to the user.

**Review by mickadoo — APPROVED** [review]
re-approving after revision changes

**Comment by eileenmcnaughton** [comment]
OK - I'm happy to go with @mickadoo's re-review on this. It seems to have been conceptually agreed & the code makes sense from a quick skim so I don't have any concerns over & above @mickadoo having tested /reviewed

File: CRM/Upgrade/Incremental/php/FiveTwo.php [hunk 4]
```diff
@@ -40,10 +40,14 @@ class CRM_Upgrade_Incremental_php_FiveTwo extends CRM_Upgrade_Incremental_Base {
    * @param null $currentVer
    */
   public function setPreUpgradeMessage(&$preUpgradeMessage, $rev, $currentVer = NULL) {
-    // Example: Generate a pre-upgrade message.
-    // if ($rev == '5.12.34') {
-    //   $preUpgradeMessage .= '<p>' . ts('A new permission has been added called %1 This Permission is now used to control access to the Manage Tags screen', array(1 => 'manage tags')) . '</p>';
-    // }
+    if ($rev == '5.3.0') {
```
Comment (seamuslee001) [inline CRM/Upgrade/Incremental/php/FiveTwo.php]: @eileenmcnaughton can you move this code into FiveThree.php

Comment (eileenmcnaughton) [reply CRM/Upgrade/Incremental/php/FiveTwo.php]: ah dang - didn't spot that!

**Comment by eileenmcnaughton** [comment]
This appears to have caused a regression - see https://github.com/civicrm/civicrm-core/pull/12896

**Pull request merged**